### PR TITLE
Avoid head-of-line blocking when handling large volumes

### PR DESCRIPTION
**Pull request by eriknordmark**
Separated across a few PRs.

This introduces a new worker.Pool which provides the same API as the existing worker, but runs separate workers in parallel based on the load. The count argument is the max number of workers it will run concurrently.
Note that should we exceed that number we explicitly log.Fatal. We could run this with an unbounded upper limit instead of 20, but that will consume more memory and potentially cause thrashing if we are doing that many volumes and/or CAS operations in parallel.

The changes to cmd/baseosmgr and cmd/volumemgr are quite small since the API (once we have the TrySubmit API) are basically identical.

### CODE REVIEW

**Review by deitch — COMMENTED** [review]
I have a few questions:

* why a new package `workerpool` instead of just a new struct inside `worker`? I think that would be cleaner / easier to understand.
* I have never loved `Try*` as a function name, but that is just preference.
* do we need a `done, err` for `TrySubmit`? I think it might be more consistent with the language if we just return an error, and have a custom error `worker.FullQueue`. We can test for that. It also simplifies the `if / else if` statements after each `TrySubmit`
* since we now have `worker.Worker` and `worker.WorkerPool` (or maybe it should just be `worker.Pool`), maybe `worker.Worker` should be an interface that both `worker.NewWorker` and `worker.NewPool` satisfies.
* can we extract the default capacity of parallel workers to a public const in the package and just use it in `NewPool`?
* how does the submit actually work? It looks like we create a slice of `worker.Worker` of capacity 10 (or whatever length) but length 0, so how does it find any to submit?

The other question I have is about the queue. Should we be creating these workers on start? Or should we be doing this lambda-style, where we create worker go routines on demand to process them until we hit capacity?

File: pkg/pillar/workerpool/workerpool.go
Comment (deitch) [inline pkg/pillar/workerpool/workerpool.go:41]: Should this slice capacity be set by `maxWorkers`?

**Comment by eriknordmark** [comment]
> * I have never loved `Try*` as a function name, but that is just preference.

Any better suggestion? I know things like mutex_enter vs. mutex_tryenter.
Feels a bit more explicit to have a separate function as opposed to a "block bool" argument since the done return value only makes sense for the try variant.

**Comment by eriknordmark** [comment]
> * do we need a `done, err` for `TrySubmit`? I think it might be more consistent with the language if we just return an error, and have a custom error `worker.FullQueue`. We can test for that. It also simplifies the `if / else if` statements after each `TrySubmit`

But semantically it isn't an error. That it what "Try" means.

Having done, err := foo() seems more clear to me that having to look at errors to say what the state is. Very different than to look an an error to tell who originated it (TLS vs. kernel TCP for instance); both of those are failures. 

This is analogous to the language starting with "item := mymap[key]" and it just panics and then realizing that "item, ok := mymap[key]" is much more useful. Sure, could just error/panic and tell from the error/panic that the key was missing but why.

**Comment by eriknordmark** [comment]
> * can we extract the default capacity of parallel workers to a public const in the package and just use it in `NewPool`?

Don't understand your suggestion.
Which default? The 20 they two callers use?
golang unlike python do not have default values for omitted parameters.

> * how does the submit actually work? It looks like we create a slice of `worker.Worker` of capacity 10 (or whatever length) but length 0, so how does it find any to submit?

First TrySubmit creates a new worker as long as it is below the limit.

**Comment by deitch** [comment]
> Any better suggestion? I know things like mutex_enter vs. mutex_tryenter.
Feels a bit more explicit to have a separate function as opposed to a "block bool" argument since the done return value only makes sense for the try variant.

I wish I did :-)

**Comment by deitch** [comment]
> Don't understand your suggestion.
> Which default? The 20 they two callers use?
> golang unlike python do not have default values for omitted parameters.

Yes. The default limit of 10 [here](https://github.com/lf-edge/eve/pull/1584/files#diff-2edd0432aadb4ebc90a35b79e14a11066e44be087dd931f4cee0e6304af9c3d5R40-R45) limit is hard-coded into the function. It would be better if it were:

```go
const (
    DefaultPoolSize = 10
)
// later
func NewWorkerPool(log worker.Logger, ctx interface{}, maxWorkers int, handlers map[string]worker.Handler) *WorkerPool {
 	workers := make([]myworker, 0, DefaultPoolSize)
	length := maxWorkers
 	if length == 0 {
 		length = DefaultPoolSize
 	}
```

**Comment by deitch** [comment]
> First TrySubmit creates a new worker as long as it is below the limit.

I missed that, which was my question. And when a worker is done, it goes back to the pool?

**Comment by eriknordmark** [comment]
> Yes. The default limit of 10 [here](https://github.com/lf-edge/eve/pull/1584/files#diff-2edd0432aadb4ebc90a35b79e14a11066e44be087dd931f4cee0e6304af9c3d5R40-R45) limit is hard-coded into the function. It would be better if it were:

That isn't a limit but just the initial capacity of the slice. As we append it will grow past that. I can get rid of that capacity - the efficency it adds would be in the noise.

**Comment by eriknordmark** [comment]
> I missed that, which was my question. And when a worker is done, it goes back to the pool?

They are always in the pool, but get skipped if they are busy i.e., if they are not blocked on reading on their request channel. 

The GC removes workers from the pool if they haven't been used for a while (and I'll fix the "while" to have a more sane value but need the tests to use a short value).

File: pkg/pillar/workerpool/workerpool.go
Comment (eriknordmark) [reply pkg/pillar/workerpool/workerpool.go:41]: No, since it is merely an initial size allocation for the array/slice. Will remove it instead.

Comment (eriknordmark) [reply pkg/pillar/workerpool/workerpool.go]: Done @deitch

File: pkg/pillar/worker/worker.go
Comment (deitch) [inline pkg/pillar/worker/worker.go:19]: We don't use this anymore, as `worker` module is a library that will return an error and possibly log a message, but never `Fatalf()`.

**Review by deitch — CHANGES_REQUESTED** [review]
This looks nice and clean (and I have no clue what most of the yetus errors mean :-) ).

The only recommended change is removing `Fatalf` from the `worker.Logger` interface, since we never `Fatalf()` inside the `worker` library.

Other than that, looks great.

File: pkg/pillar/worker/worker.go
Comment (eriknordmark) [reply pkg/pillar/worker/worker.go:19]: Good catch

**Review by deitch — APPROVED** [review]
lgtm

**Comment by rvs** [comment]
This looks OK, but one statement from @deitch  caught my eye:

> and I have no clue what most of the yetus errors mean

I don't think we can have that -- I think we need to have a pretty clear policy on when we merge (otherwise it is going to be very difficult to stay on the same page). Now, granted we have a brand new CI/CD with brand new checks in place -- so we have to agree on what is and what isn't part of the criteria. But we have to do it explicitly and pretty fast.

WDYT @eriknordmark @deitch -- what's the best forum to hammer out that agreement?

**Comment by deitch** [comment]
Unfortunately, I cannot find the output again. It was built in here before but now that it is merged, I cannot find it. Let's find it, post it to an issue, and then discuss it in that issue.

**Comment by rvs** [comment]
It is actually right there if click on top-level "Files Changed" tab in this very PR and keep scrolling.

**Comment by deitch** [comment]
Oh yeah, I see it now. I thought it would be in actions.

**Pull request merged**
